File: cores/csharp/src/Cartridge.cs
using System.Runtime.CompilerServices;

namespace NesCore;

// iNES cartridge loader + mapper dispatch (port of src/cartridge.rs).
public sealed class Cartridge
{
    public const int HeaderSize = 16;
    public const int TrainerSize = 512;
    public const uint PrgRomUnit = 16384;
    public const uint ChrRomUnit = 8192;

    private static readonly byte[] InesMagic = { (byte)'N', (byte)'E', (byte)'S', 0x1A };
    private static readonly byte[] FdsMagic = { (byte)'F', (byte)'D', (byte)'S', 0x1A };
    private const uint FdsHeaderSize = 16;
    private const uint FdsDiskSideSize = 65500;

    public InesHeader Header;
    public Mapper Mapper;

    public Cartridge() { }

    public static int ParseHeader(byte[] bytes, int len, out InesHeader hdr)
    {
        hdr = default;
        if (len < HeaderSize) return 1;
        for (int i = 0; i < 4; ++i)
            if (bytes[i] != InesMagic[i]) return 2;

        byte prgRomBanks = bytes[4];
        byte chrRomBanks = bytes[5];
        byte flags6 = bytes[6];
        byte flags7 = bytes[7];

        bool hasTrainer = (flags6 & 0x04) != 0;
        bool hasBattery = (flags6 & 0x02) != 0;
        bool fourScreen = (flags6 & 0x08) != 0;
        bool vertical = (flags6 & 0x01) != 0;

        Mirroring mirroring;
        if (fourScreen) mirroring = Mirroring.FourScreen;
        else if (vertical) mirroring = Mirroring.Vertical;
        else mirroring = Mirroring.Horizontal;

        ushort mapperNumber = (ushort)(((ushort)(flags6 >> 4)) | (ushort)(((ushort)(flags7 >> 4)) << 4));
        byte tvSystem = (byte)(bytes[9] & 0x03);

        hdr.PrgRomBanks = prgRomBanks;
        hdr.ChrRomBanks = chrRomBanks;
        hdr.MapperNumber = mapperNumber;
        hdr.Mirroring = mirroring;
        hdr.HasTrainer = hasTrainer;
        hdr.HasBattery = hasBattery;
        hdr.TvSystem = tvSystem;
        return 0;
    }

    public static int FromBytes(byte[] bytes, int len, out Cartridge cart)
    {
        cart = new Cartridge();
        int rc = ParseHeader(bytes, len, out InesHeader hdr);
        if (rc != 0) return rc;

        uint prgSize = (uint)hdr.PrgRomBanks * PrgRomUnit;
        int prgOff = HeaderSize;
        if (hdr.HasTrainer) prgOff += TrainerSize;
        uint chrSize = (hdr.ChrRomBanks > 0) ? (uint)hdr.ChrRomBanks * ChrRomUnit : 0u;
        if (len < prgOff + prgSize + chrSize) return 5;

        byte[] prgRom = new byte[prgSize];
        System.Array.Copy(bytes, prgOff, prgRom, 0, (int)prgSize);
        byte[] chrRom = (chrSize > 0) ? new byte[chrSize] : null;
        if (chrRom != null)
            System.Array.Copy(bytes, prgOff + (int)prgSize, chrRom, 0, (int)chrSize);

        cart.Header = hdr;
        cart.Mapper = MapperFactory.Create(hdr.MapperNumber, prgRom, prgSize, chrRom, chrSize,
                                            hdr.Mirroring, hdr.HasBattery);
        return 0;
    }

    public static int FromFdsBytes(byte[] diskData, int diskLen, byte[] bios, int biosLen, out Cartridge cart)
    {
        cart = new Cartridge();
        if (diskLen < FdsHeaderSize) return 1;
        for (int i = 0; i < 4; ++i)
            if (diskData[i] != FdsMagic[i]) return 2;
        byte diskCount = diskData[4];
        if (diskCount == 0) return 6;
        uint needed = FdsHeaderSize + (uint)diskCount * FdsDiskSideSize;
        if (diskLen < needed) return 1;

        uint rawSize = (uint)diskCount * FdsDiskSideSize;
        byte[] rawDisk = new byte[rawSize];
        System.Array.Copy(diskData, (int)FdsHeaderSize, rawDisk, 0, (int)rawSize);

        cart.Header = new InesHeader
        {
            MapperNumber = 20,
            Mirroring = Mirroring.Vertical,
            PrgRomBanks = 0,
            ChrRomBanks = 0,
        };
        cart.Mapper = new Fds(bios, biosLen, rawDisk, rawSize);
        return 0;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public byte ReadPrg(ushort addr) => Mapper.ReadPrg(addr);
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public byte ReadPrgMut(ushort addr) => Mapper.ReadPrgMut(addr);
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void WritePrg(ushort addr, byte value) => Mapper.WritePrg(addr, value);
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public byte ReadChr(ushort addr) => Mapper.ReadChr(addr);
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public byte ReadChrLatched(ushort addr) => Mapper.ReadChrLatched(addr);
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void WriteChr(ushort addr, byte value) => Mapper.WriteChr(addr, value);
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public Mirroring MirrorMode() => Mapper.MirrorMode();
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool ChrIsRam() => Mapper.ChrIsRam();
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool HasBattery() => Mapper.HasBattery();
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool IrqPending() => Mapper.IrqPending();
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void ClockIrq() => Mapper.ClockIrq();
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void ResetScanlineCounter() => Mapper.ResetScanlineCounter();
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void ClockCpu(uint cpuCycles) => Mapper.ClockCpu(cpuCycles);
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public float ExpansionAudioSample() => Mapper.ExpansionAudioSample();
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public uint SaveState(byte[] buf) => Mapper.SaveState(buf);
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool LoadState(byte[] buf, uint len) => Mapper.LoadState(buf, len);
}

File: cores/typescript/src/cartridge.ts
// iNES cartridge loader + mapper dispatch (port of cores/c/src/cartridge.c).

import { InesHeader, makeInesHeader, Mapper, Mirroring, createMapper } from './mapper';

export class Cartridge {
    static readonly HeaderSize = 16;
    static readonly TrainerSize = 512;
    static readonly PrgRomUnit = 16384;
    static readonly ChrRomUnit = 8192;

    static readonly InesMagic = [0x4E, 0x45, 0x53, 0x1A]; // "NES\x1A"

    header: InesHeader = makeInesHeader();
    mapper!: Mapper;

    static parseHeader(bytes: Uint8Array, len: number, hdr: InesHeader): number {
        if (len < Cartridge.HeaderSize) return 1;
        for (let i = 0; i < 4; ++i)
            if (bytes[i] !== Cartridge.InesMagic[i]) return 2;

        const prgRomBanks = bytes[4];
        const chrRomBanks = bytes[5];
        const flags6 = bytes[6];
        const flags7 = bytes[7];

        const hasTrainer = (flags6 & 0x04) !== 0;
        const hasBattery = (flags6 & 0x02) !== 0;
        const fourScreen = (flags6 & 0x08) !== 0;
        const vertical = (flags6 & 0x01) !== 0;

        let mirroring: number;
        if (fourScreen) mirroring = Mirroring.FourScreen;
        else if (vertical) mirroring = Mirroring.Vertical;
        else mirroring = Mirroring.Horizontal;

        const mapperNumber = ((flags6 >> 4) | ((flags7 >> 4) << 4)) & 0xFFFF;
        const tvSystem = bytes[9] & 0x03;

        hdr.prgRomBanks = prgRomBanks;
        hdr.chrRomBanks = chrRomBanks;
        hdr.mapperNumber = mapperNumber;
        hdr.mirroring = mirroring;
        hdr.hasTrainer = hasTrainer;
        hdr.hasBattery = hasBattery;
        hdr.tvSystem = tvSystem;
        return 0;
    }

    static fromBytes(bytes: Uint8Array, len: number): Cartridge | null {
        const cart = new Cartridge();
        const hdr = makeInesHeader();
        const rc = Cartridge.parseHeader(bytes, len, hdr);
        if (rc !== 0) return null;

        const prgSize = hdr.prgRomBanks * Cartridge.PrgRomUnit;
        let prgOff = Cartridge.HeaderSize;
        if (hdr.hasTrainer) prgOff += Cartridge.TrainerSize;
        const chrSize = (hdr.chrRomBanks > 0) ? hdr.chrRomBanks * Cartridge.ChrRomUnit : 0;
        if (len < prgOff + prgSize + chrSize) return null;

        const prgRom = new Uint8Array(prgSize);
        prgRom.set(bytes.subarray(prgOff, prgOff + prgSize));
        let chrRom: Uint8Array | null = null;
        if (chrSize > 0) {
            chrRom = new Uint8Array(chrSize);
            chrRom.set(bytes.subarray(prgOff + prgSize, prgOff + prgSize + chrSize));
        }

        cart.header = hdr;
        cart.mapper = createMapper(hdr.mapperNumber, prgRom, prgSize, chrRom, chrSize, hdr.mirroring, hdr.hasBattery);
        return cart;
    }

    readPrg(addr: number): number { return this.mapper.readPrg(addr); }
    readPrgMut(addr: number): number { return this.mapper.readPrgMut(addr); }
    writePrg(addr: number, value: number): void { this.mapper.writePrg(addr, value); }
    readChr(addr: number): number { return this.mapper.readChr(addr); }
    readChrLatched(addr: number): number { return this.mapper.readChrLatched(addr); }
    writeChr(addr: number, value: number): void { this.mapper.writeChr(addr, value); }
    mirrorMode(): number { return this.mapper.mirrorMode(); }
    chrIsRam(): boolean { return this.mapper.chrIsRam(); }
    hasBattery(): boolean { return this.mapper.hasBattery(); }
    irqPending(): boolean { return this.mapper.irqPending(); }
    clockIrq(): void { this.mapper.clockIrq(); }
    resetScanlineCounter(): void { this.mapper.resetScanlineCounter(); }
    clockCpu(cpuCycles: number): void { this.mapper.clockCpu(cpuCycles); }
    expansionAudioSample(): number { return this.mapper.expansionAudioSample(); }
    saveState(buf: Uint8Array | null): number { return this.mapper.saveState(buf); }
    loadState(buf: Uint8Array, len: number): boolean { return this.mapper.loadState(buf, len); }
}
